cpp/src/knn/knn.cu

```diff
@@ -49,24 +50,108 @@ void brute_force_knn(const raft::handle_t& handle,
+    // for offsets of the indices, build them
+    // from the local partitions
+    id_ranges       = new std::vector<int64_t>();
+    int64_t total_n = 0;
+    for (size_t i = 0; i < input.size(); i++) {
+      id_ranges->push_back(total_n);
+      total_n += sizes[i];
+    }
+  } else {
+    // otherwise, use the given translations
+    id_ranges = translations;
+  }
+
+  rmm::device_uvector<int64_t> trans(id_ranges->size(), userStream);
+  raft::update_device(trans.data(), id_ranges->data(), id_ranges->size(), userStream);
+
+  rmm::device_uvector<float> all_D(0, userStream);
+  rmm::device_uvector<int64_t> all_I(0, userStream);
+
+  float* out_D   = res_D;
+  int64_t* out_I = res_I;
+
+  if (input.size() > 1) {
+    all_D.resize(input.size() * k * n, userStream);
+    all_I.resize(input.size() * k * n, userStream);
+
+    out_D = all_D.data();
+    out_I = all_I.data();
+  }
+
+  // Make other streams from pool wait on main stream
+  raft::resource::wait_stream_pool_on_stream(handle);
+
+  for (size_t i = 0; i < input.size(); i++) {
+    float* out_d_ptr   = out_D + (i * k * n);
+    int64_t* out_i_ptr = out_I + (i * k * n);
+
+    auto stream         = raft::resource::get_next_usable_stream(handle, i);
+    auto current_handle = raft::device_resources(stream);
+
+    // build the brute_force index (precalculates norms etc)
+    std::optional<cuvs::neighbors::brute_force::index<float>> idx;
+    if (rowMajorIndex) {
+      idx = cuvs::neighbors::brute_force::build(
+        current_handle,
+        raft::make_device_matrix_view<const float, int64_t, raft::row_major>(input[i], sizes[i], D),
+        static_cast<cuvs::distance::DistanceType>(metric),
+        metric_arg);
+
+    } else {
+      idx = cuvs::neighbors::brute_force::build(
+        current_handle,
+        raft::make_device_matrix_view<const float, int64_t, raft::col_major>(input[i], sizes[i], D),
+        static_cast<cuvs::distance::DistanceType>(metric),
+        metric_arg);
+    }
+
+    // query the index
+    if (rowMajorQuery) {
+      cuvs::neighbors::brute_force::search(
+        current_handle,
+        *idx,
+        raft::make_device_matrix_view<const float, int64_t, raft::row_major>(search_items, n, D),
+        raft::make_device_matrix_view<int64_t, int64_t>(out_i_ptr, n, k),
+        raft::make_device_matrix_view<float, int64_t>(out_d_ptr, n, k));
+    } else {
+      cuvs::neighbors::brute_force::search(
+        current_handle,
+        *idx,
+        raft::make_device_matrix_view<const float, int64_t, raft::col_major>(search_items, n, D),
+        raft::make_device_matrix_view<int64_t, int64_t>(out_i_ptr, n, k),
+        raft::make_device_matrix_view<float, int64_t>(out_d_ptr, n, k));
+    }
+  }
+
+  // Sync internal streams if used. We don't need to
+  // sync the user stream because we'll already have
+  // fully serial execution.
+  raft::resource::sync_stream_pool(handle);
+
+  if (input.size() > 1 || translations != nullptr) {
```

Comment:
In the most ideal case, we should be able to formalize an "id_mapper" API at some point and have it used consistently for all index types in cuVS itself. That's a pretty trivial gather operation that could be accelerated.